src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -0,0 +1,1597 @@
+  sendvec = Array<dcomplex>(nmode);
+  recvecin = Array<dcomplex>(nmode);
+  recvecout = Array<dcomplex>(nmode);
+  MPI_Request rreqin, rreqout;
+
+  // Included processors send their contribution to procs that are included on
+  // the level above.
+  // Special case: last proc sends if on level > 1, but NOT on level 1
+  if(l.included and (not localmesh->lastX() or l.current_level>1) ){
+    for(int kz=0; kz<nmode; kz++){
+      if(!converged[kz]){
+	fine_error(1,kz) = l.xloc(1,kz);
+	sendvec[kz] = l.xloc(1,kz);
+	if(localmesh->lastX()){
+	  fine_error(2,kz) = l.xloc(2,kz);
+	}
+      }
+    }
+
+    if(!localmesh->lastX()){
+      MPI_Send(&sendvec[0], nmode, MPI_DOUBLE_COMPLEX, lup.proc_out, 0, BoutComm::get());
+    }
+    if(!localmesh->firstX()){
+      MPI_Send(&sendvec[0], nmode, MPI_DOUBLE_COMPLEX, lup.proc_in, 1, BoutComm::get());
+    }
+
+  }
+
+  // Receive if proc is included on the level above, but not this level.
+  // Special case: last proc receives if on level 1
+  if( (l.included_up and not l.included) or (localmesh->lastX() and l.current_level==1) ){
+    if(!localmesh->firstX()){
+      MPI_Irecv(&recvecin[0], nmode, MPI_DOUBLE_COMPLEX, lup.proc_in, 0, BoutComm::get(), &rreqin);
+    }
+    if(!localmesh->lastX()){
+      MPI_Irecv(&recvecout[0], nmode, MPI_DOUBLE_COMPLEX, lup.proc_out, 1, BoutComm::get(), &rreqout);
+    }
+
+    for(int kz=0; kz < nmode; kz++){
+      fine_error(1,kz) = 0.0;
+    }
+
+    if(!localmesh->firstX()){
+      MPI_Wait(&rreqin,MPI_STATUS_IGNORE);
+      for(int kz=0; kz < nmode; kz++){
+	if(!converged[kz]){
+	  fine_error(1,kz) += 0.5*recvecin[kz];
+	}
+      }
+    }
+    if(!localmesh->lastX()){
+      MPI_Wait(&rreqout,MPI_STATUS_IGNORE);
+      for(int kz=0; kz < nmode; kz++){
+	if(!converged[kz]){
+	  fine_error(1,kz) += 0.5*recvecout[kz];
+	}
+      }
+    }
+  }
+  // Special case where we need to fill (1,kz) on final proc
+  if(localmesh->lastX() and l.current_level==1){
+    for(int kz=0; kz < nmode; kz++){
+      if(!converged[kz]){
+	fine_error(1,kz) += 0.5*l.xloc(2,kz);
+      }
+    }
+  }
+}
+
+/*
+ * Synchronize the values of a reduced field(4,nmode) between processors that
+ * are neighbours on level l. This assumes each processor's value of
+ * field(1,:) is correct, and puts the in-neighbour's value into field(0,:)
+ * and out-neighbour's value into field(3,:).
+ */
+void LaplaceIPT::synchronize_reduced_field(const Level &l, Matrix<dcomplex> &field){
+
+  SCOREP0();
+  if(l.included){
+
+    MPI_Comm comm = BoutComm::get();
+    int err;
+
+    // Communicate in
+    if(!localmesh->firstX()){
+      err = MPI_Sendrecv(&field(1,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_in, 1, &field(0,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_in, 0, comm, MPI_STATUS_IGNORE);
+    }
+
+    // Communicate out
+    if(!localmesh->lastX()){
+      err = MPI_Sendrecv(&field(1,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_out, 0, &field(3,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_out, 1, comm, MPI_STATUS_IGNORE);
```

Comment:
Oh, no, that's a bug.  The finest grid is index 1 on every processor, and index 2 on the last processor.  This means we want to fill neighbours' guard cells with my index 1 most of the time.  However on all coarse grids (level > 0), on the last proc index 1 is not included and we only have index 2.  We therefore need to send index 2 down for (lastX and level>0), but index 1 otherwise.

I've corrected it.  (It doesn't change the result though! It was correct for the finest level, which is the most important thing.  Being wrong on coarser levels didn't seem to slow convergence either.  But still, better to be correct!)